### PR TITLE
introduce remote-tests for pallet-bags-list

### PR DESCRIPTION
**Pull request by kianenigma**
This brings over a collection of cool test for the pallet-bags-list to substrate. Some of these tests used to live in polkadot repo and are mostly copy-pasted. Some are new. In principle, they should be here since pallet-bags-list can live in any runtime. 

Note that the remote tests crate is a library crate that can be easily wrapper in `#[tokio::test]` or a `fn main() {}`. See the companion as an example. 

polkadot companion: https://github.com/paritytech/polkadot/pull/4065

### CODE REVIEW

File: frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -668,7 +668,8 @@ impl<T: Config> Pallet<T> {
 			let nominator_count = CounterForNominators::<T>::get() as usize;
 			let validator_count = CounterForValidators::<T>::get() as usize;
 			let all_voter_count = validator_count.saturating_add(nominator_count);
-			maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count)
+			maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count) // TODO: this code is not sensible
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:671]: you are right; the `max_max_len` is coming from a source where we assume it know the memory limits. In reality, it is `Some(22500)`.

My comment about the code being insensible was wrong.

File: frame/bags-list/remote-tests/src/migration.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			pallets: vec![<Runtime as frame_system::Config>::PalletInfo::name::<
+				pallet_staking::Pallet<Runtime>,
+			>()
+			.expect("Pallet always has name; qed.")
+			.to_string()],
+			at: None,
+			state_snapshot: None,
+		}))
+		.inject_hashed_prefix(&<pallet_staking::Bonded<Runtime>>::prefix_hash())
+		.inject_hashed_prefix(&<pallet_staking::Ledger<Runtime>>::prefix_hash())
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/bags-list/remote-tests/src/migration.rs:41]: yeah good point.

File: utils/frame/remote-externalities/src/lib.rs [hunk 3]
```diff
@@ -397,7 +397,7 @@ impl<B: BlockT> Builder<B> {
 
 		for key in &self.hashed_keys {
 			let key = StorageKey(key.to_vec());
-			debug!(target: LOG_TARGET, "adding data for hashed key: {:?}", HexDisplay::from(&key));
+			info!(target: LOG_TARGET, "adding data for hashed key: {:?}", HexDisplay::from(&key));
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply utils/frame/remote-externalities/src/lib.rs:400]: yeah, intentional.

File: frame/bags-list/src/lib.rs [hunk 4]
```diff
@@ -155,7 +153,7 @@ pub mod pallet {
 
 	/// How many ids are registered.
 	// NOTE: This is merely a counter for `ListNodes`. It should someday be replaced by the
-	// `CountedMaop` storage.
+	// `CountedMap` storage.
 	#[pallet::storage]
```
Comment (thiolliere) [inline frame/bags-list/src/lib.rs:157]: sidenote: counted map is now merged in master

**Review by thiolliere — APPROVED** [review]

File: frame/bags-list/src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply frame/bags-list/src/lib.rs:157]: yeah I have a self-note to move this and the staking counters to this new type. Neither should require any migration.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**